lib/mongoose_push/service/apns/supervisor.ex

```diff
@@ -72,4 +98,69 @@ defmodule MongoosePush.Service.APNS.Supervisor do
 
     [single_config | mode_list]
   end
+
+  defp convert_token_pool_to_sparrow({pool_name, pool_config}, {mode_list, tokens}) do
+    auth_type = {:auth_type, :token_based}
+    pool_size = {:worker_num, pool_config[:pool_size]}
+
+    endpoint_mode = @default_endpoints[pool_config[:mode]]
+    endpoint = {:endpoint, pool_config[:endpoint] || endpoint_mode}
+
+    port_config =
+      case pool_config[:use_2197] do
+        true ->
+          2197
+
+        _ ->
+          nil
+      end
+
+    port = {:port, port_config}
+
+    name = {:pool_name, pool_name}
+
+    key = pool_config[:auth][:key_id]
+    team = pool_config[:auth][:team_id]
+    p8_file_path = pool_config[:auth][:p8_file_path]
+
+    if not File.exists?(p8_file_path) do
+      Logger.error(~s"Required file does not exist in the given path:
+      p8_file=#{p8_file_path}")
+
+      Supervisor.stop(self(), {:error, "Required file does not exist in the given path:
+      p8_file=#{p8_file_path}"})
+    end
+
+    if is_nil(key) or is_nil(team) do
+      Logger.error(~s"Required authentication elements are missing:
+      key=#{key} or team=#{team}")
+
+      Supervisor.stop(self(), {:error, "Required file does not exist in the given path:
+      p8_file=#{p8_file_path}"})
+    end
+
+    key_id = {:key_id, key}
+    team_id = {:team_id, team}
+    p8_file = {:p8_file_path, p8_file_path}
+    token_id = random_atom(15)
+    pool_token = {:token_id, token_id}
+
+    single_config =
+      [auth_type, pool_token, pool_size, endpoint, port, name]
+      |> Enum.filter(fn {_key, value} -> !is_nil(value) end)
+
+    token = [pool_token, key_id, team_id, p8_file]
```

Comment:
Maybe provide Struct for the token ??

Reply:
This looks like a constant size list, which is just weird. Only after analyzing the whole function, I noticed, that all elements are 2-tuples, which makes it a keyword list, which is fine. IMHO, this keyword list, should be created here, inline, so that all keys are listed in one place, explicitly and only values are referenced by a variable.